util/jsonutil/merge.go

```diff
@@ -0,0 +1,183 @@
+func MergeClone(v any, data json.RawMessage) error {
+	err := jsonConfigMergeClone.Unmarshal(data, v)
+	if err != nil {
+		return &errortypes.FailedToUnmarshal{
+			Message: tryExtractErrorMessage(err),
+		}
+	}
+	return err
+}
+
+type mergeCloneExtension struct {
+	jsoniter.DummyExtension
+}
+
+func (e *mergeCloneExtension) CreateDecoder(typ reflect2.Type) jsoniter.ValDecoder {
+	if typ == jsonRawMessageType {
+		return &extMergeDecoder{sliceType: typ.(*reflect2.UnsafeSliceType)}
+	}
+	return nil
+}
+
+func (e *mergeCloneExtension) DecorateDecoder(typ reflect2.Type, decoder jsoniter.ValDecoder) jsoniter.ValDecoder {
+	if typ.Kind() == reflect.Ptr {
+		ptrType := typ.(*reflect2.UnsafePtrType)
+		return &ptrCloneDecoder{valueDecoder: decoder, elemType: ptrType.Elem()}
+	}
+
+	// don't use json.RawMessage on fields handled by extMergeDecoder
+	if typ.Kind() == reflect.Slice && typ != jsonRawMessageType {
+		return &sliceCloneDecoder{valueDecoder: decoder, sliceType: typ.(*reflect2.UnsafeSliceType)}
+	}
+
+	if typ.Kind() == reflect.Map {
+		return &mapCloneDecoder{valueDecoder: decoder, mapType: typ.(*reflect2.UnsafeMapType)}
+	}
+
+	return decoder
+}
+
+type ptrCloneDecoder struct {
+	elemType     reflect2.Type
+	valueDecoder jsoniter.ValDecoder
+}
+
+func (d *ptrCloneDecoder) Decode(ptr unsafe.Pointer, iter *jsoniter.Iterator) {
+	// don't clone if field is being set to nil. checking for nil "consumes" the null
+	// token, so must be handled in this decoder.
+	if iter.ReadNil() {
+		*((*unsafe.Pointer)(ptr)) = nil
+		return
+	}
+
+	// clone if there is an existing object. creation of new objects is handled by the
+	// original decoder.
+	if *((*unsafe.Pointer)(ptr)) != nil {
+		obj := d.elemType.UnsafeNew()
+		d.elemType.UnsafeSet(obj, *((*unsafe.Pointer)(ptr)))
+		*((*unsafe.Pointer)(ptr)) = obj
+	}
+
+	d.valueDecoder.Decode(ptr, iter)
+}
+
+type sliceCloneDecoder struct {
+	sliceType    *reflect2.UnsafeSliceType
+	valueDecoder jsoniter.ValDecoder
+}
+
+func (d *sliceCloneDecoder) Decode(ptr unsafe.Pointer, iter *jsoniter.Iterator) {
+	// don't clone if field is being set to nil. checking for nil "consumes" the null
+	// token, so must be handled in this decoder.
+	if iter.ReadNil() {
+		d.sliceType.UnsafeSetNil(ptr)
+		return
+	}
+
+	// clone if there is an existing object. creation of new objects is handled by the
+	// original decoder.
+	if !d.sliceType.UnsafeIsNil(ptr) {
+		clone := reflectutil.UnsafeSliceClone(ptr, d.sliceType)
+		d.sliceType.UnsafeSet(ptr, clone)
+	}
+
+	d.valueDecoder.Decode(ptr, iter)
+}
+
+type mapCloneDecoder struct {
+	mapType      *reflect2.UnsafeMapType
+	valueDecoder jsoniter.ValDecoder
+}
+
+func (d *mapCloneDecoder) Decode(ptr unsafe.Pointer, iter *jsoniter.Iterator) {
+	// don't clone if field is being set to nil. checking for nil "consumes" the null
+	// token, so must be handled in this decoder.
+	if iter.ReadNil() {
+		*(*unsafe.Pointer)(ptr) = nil
+		d.mapType.UnsafeSet(ptr, d.mapType.UnsafeNew())
```

Comment:
Is this call initializing `ptr` to a new empty map? If so, slices and pointers in lines 80 and 104 are simply set to `nil`.

Reply:
I do not fully understand why setting the pointer to `nil` isn't sufficient for maps. This is copy I copied/pasted from json-iter, see: https://github.com/json-iterator/go/blob/71ac16282d122fdd1e3a6d3e7f79b79b4cc3b50e/reflect_map.go#L158.